From: Igor Maravic <igorm@etf.rs>
Date: Thu, 1 Dec 2011 23:48:20 +0000
Subject: forcedeath: Fix bql support for forcedeath

commit 7505afe28c16a8d386624930a018d0052c75d687 upstream.

Moved netdev_completed_queue() out of while loop in function nv_tx_done_optimized().
Because this function was in while loop,
BUG_ON(count > dql->num_queued - dql->num_completed)
was hit in dql_completed().

Signed-off-by: Igor Maravic <igorm@etf.rs>
Acked-by: Eric Dumazet <eric.dumazet@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/ethernet/nvidia/forcedeth.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: linux-3.2.46/drivers/net/ethernet/nvidia/forcedeth.c
===================================================================
--- linux-3.2.46.orig/drivers/net/ethernet/nvidia/forcedeth.c	2013-07-26 19:24:30.000000000 +0000
+++ linux-3.2.46/drivers/net/ethernet/nvidia/forcedeth.c	2013-07-26 19:24:31.000000000 +0000
@@ -2470,13 +2470,14 @@
 				nv_tx_flip_ownership(dev);
 		}
 
-		netdev_completed_queue(np->dev, tx_work, bytes_cleaned);
-
 		if (unlikely(np->get_tx.ex++ == np->last_tx.ex))
 			np->get_tx.ex = np->first_tx.ex;
 		if (unlikely(np->get_tx_ctx++ == np->last_tx_ctx))
 			np->get_tx_ctx = np->first_tx_ctx;
 	}
+
+	netdev_completed_queue(np->dev, tx_work, bytes_cleaned);
+
 	if (unlikely((np->tx_stop == 1) && (np->get_tx.ex != orig_get_tx))) {
 		np->tx_stop = 0;
 		netif_wake_queue(dev);
